cmd/mattermost/commands/plugin.go

```diff
@@ -0,0 +1,185 @@
+	RunE:    pluginDisableCmdF,
+}
+
+var PluginListCmd = &cobra.Command{
+	Use:     "list",
+	Short:   "List plugins",
+	Long:    "List all active and inactive plugins installed on your Mattermost server.",
+	Example: `  plugin list`,
+	RunE:    pluginListCmdF,
+}
+
+func init() {
+	PluginCmd.AddCommand(
+		PluginAddCmd,
+		PluginDeleteCmd,
+		PluginEnableCmd,
+		PluginDisableCmd,
+		PluginListCmd,
+	)
+	RootCmd.AddCommand(PluginCmd)
+}
+
+func pluginAddCmdF(command *cobra.Command, args []string) error {
+	a, err := InitDBCommandContextCobra(command)
+	if err != nil {
+		return err
+	}
+	defer a.Shutdown()
+
+	if len(args) < 1 {
+		return errors.New("Expected at least one argument. See help text for details.")
+	}
+
+	for i, plugin := range args {
+		fileReader, err := os.Open(plugin)
+		if err != nil {
+			return err
+		}
+
+		if _, err := a.InstallPlugin(fileReader); err != nil {
+			CommandPrintErrorln("Unable to add plugin: " + args[i] + ". Error: " + err.Error())
+		} else {
+			CommandPrettyPrintln("Added plugin: " + plugin)
+		}
+		fileReader.Close()
+	}
+
+	return nil
+}
+
+func pluginDeleteCmdF(command *cobra.Command, args []string) error {
+	a, err := InitDBCommandContextCobra(command)
+	if err != nil {
+		return err
+	}
+	defer a.Shutdown()
+
+	if len(args) < 1 {
+		return errors.New("Expected at least one argument. See help text for details.")
+	}
+
+	for _, plugin := range args {
+		if err := a.RemovePlugin(plugin); err != nil {
+			CommandPrintErrorln("Unable to delete plugin: " + plugin + ". Error: " + err.Error())
+		} else {
+			CommandPrettyPrintln("Deleted plugin: " + plugin)
+		}
+	}
+
+	return nil
+}
+
+func pluginEnableCmdF(command *cobra.Command, args []string) error {
+	a, err := InitDBCommandContextCobra(command)
+	if err != nil {
+		return err
+	}
+	defer a.Shutdown()
+
+	if len(args) < 1 {
+		return errors.New("Expected at least one argument. See help text for details.")
+	}
+
+	for _, plugin := range args {
+		if err := a.EnablePlugin(plugin); err != nil {
+			CommandPrintErrorln("Unable to enable plugin: " + plugin + ". Error: " + err.Error())
+		} else {
+			CommandPrettyPrintln("Enabled plugin: " + plugin)
+		}
+	}
+
+	return nil
+}
+
+func pluginDisableCmdF(command *cobra.Command, args []string) error {
+	a, err := InitDBCommandContextCobra(command)
+	if err != nil {
+		return err
+	}
+	defer a.Shutdown()
+
+	if len(args) < 1 {
+		return errors.New("Expected at least one argument. See help text for details.")
+	}
+
+	for _, plugin := range args {
+		if err := a.DisablePlugin(plugin); err != nil {
+			CommandPrintErrorln("Unable to disable plugin: " + plugin + ". Error: " + err.Error())
+		} else {
+			CommandPrettyPrintln("Disabled plugin: " + plugin)
+		}
+	}
+
+	return nil
+}
+
+func pluginListCmdF(command *cobra.Command, args []string) error {
+	a, err := InitDBCommandContextCobra(command)
+	if err != nil {
+		return err
+	}
+	defer a.Shutdown()
+
+	if pluginsResp, err := a.GetPlugins(); err != nil {
```

Comment:
I prefer the happy path in the first level of indentation, but you have the problem with the nil pointer on err.Error().

The problem was `InitDBCommandContextCobra(command)` returns an app and an `error` (of type `error` not `*model.AppError`), so you set the err variable to type `error`, after that, when you call `a.GetPlugins()` you are returnting the plugins list and a `*model.AppError`, because err already exists and is a `error` type, it doesn't redefine the error as a `*model.AppError` so the if works incorrectly.
My proposal would be, back to the previous version and replace `err` variable name with `appErr`, like this:
```go
pluginsResp, appErr := a.GetPlugins()
if appErr != nil {
    return errors.New("Unable to list plugins. Error: " + appErr.Error())
}
```

Reply:
I did not know that. I'll make sure to look at the return types for errors next time. Thanks @jespino!

Reply:
:smiley: